### DISCUSSION TITLE
More Pandas

### DISCUSSION THREAD

**Discussion by chris-ha458** [ChenghaoMou/text-dedup#55] [Ideas]
In this repo, we do a lot of string manipulation and then numerical calculations
For the former we rely mostly on python types and data structures and for the latter, numpy.
(there are some interlap however)

I feel that for simple string manipulations python works well, but for mass string manipulation, it is a bottleneck.
Numpy is not well suited to accelerate string manipulations either.
I am thinking of utilizing pandas for this step.

Pros : might be faster and depending on implementation, cleaner code.
Cons : might introduce more complexity for little gain.
Neutral : won't introduce further dependencies since pandas is already used here.

**Comment by ChenghaoMou** [comment]
Can you give an example of such optimization? 

Most of the string manipulation (tokenization) happens during the fingerprint calculation, so I don't think converting the arrow dataset first to a pandas dataframe and then performing string operations will gain much improvement unless it is some kind of batched processing where the overhead is marginalized.

**Comment by chris-ha458** [comment]
My plan was to use dataframes to enable array string processing which would enable batch in `ds.map`. 

I was under the impression that using zerocopy conversions with pandas would be "overhead free" 
If not i would just operate on the underlying arrow datasets.

I am reevaluating these underlying assumptions as they don't seem that straightforward anymore.
As for where I would use it, string manipulation was one of the places I had in mind.

**Comment by ChenghaoMou** [comment]
Each batch becomes a column-oriented dict, if I remember correctly. One needs to convert them to pandas first for each batch:
```
def process(batch: Dict[str, List[Any]]):
    df = pd.DataFrame(batch)
    ...
```

I think I tried this before because I wanted to use `groupby`, but it didn't scale well in my case.

My guess is that it will be much more fruitful to write the fingerprinting part in Cython.

**Comment by chris-ha458** [comment]
> My guess is that it will be much more fruitful to write the fingerprinting part in Cython.

It seems like that is the path 
https://github.com/mattilyra/LSH
took.

but even then they use a for loop within the cython function.
